AK/Coroutine.h

```diff
@@ -0,0 +1,213 @@
+
+    AK_MAKE_NONCOPYABLE(Coroutine);
+
+public:
+    using ReturnType = T;
+    using promise_type = Conditional<SameAs<T, void>, CoroutinePromiseVoid, CoroutinePromiseValue>;
+
+    ~Coroutine()
+    {
+        VERIFY(await_ready());
+        if constexpr (!SameAs<T, void>)
+            return_value()->~T();
+    }
+
+    Coroutine(Coroutine&& other)
+    {
+        m_handle = exchange(other.m_handle, {});
+        if (!await_ready())
+            m_handle.promise().m_coroutine = this;
+        else
+            new (return_value()) T(move(*other.return_value()));
+    }
+
+    Coroutine& operator=(Coroutine&& other)
+    {
+        if (this != &other) {
+            this->~Coroutine();
+            new (this) Coroutine(move(other));
+        }
+        return *this;
+    }
+
+    bool await_ready() const
+    {
+        return !m_handle;
+    }
+
+    void await_suspend(std::coroutine_handle<> awaiter)
+    {
+        m_handle.promise().m_awaiter = awaiter;
+    }
+
+    // Do NOT bind the result of await_resume() on a temporary coroutine (or the result of CO_TRY) to auto&&!
+    [[nodiscard]] decltype(auto) await_resume()
+    {
+        if constexpr (SameAs<T, void>)
+            return;
+        else
+            return static_cast<T&&>(*return_value());
+    }
+
+private:
+    template<typename U>
+    friend struct Detail::TryAwaiter;
+
+    // You cannot just have return_value and return_void defined in the same promise type because C++.
+    struct CoroutinePromiseBase {
+        CoroutinePromiseBase() = default;
+
+        Coroutine get_return_object()
+        {
+            return { std::coroutine_handle<promise_type>::from_promise(*static_cast<promise_type*>(this)) };
+        }
+
+        Detail::SuspendNever initial_suspend() { return {}; }
+
+        Detail::SuspendNever final_suspend() noexcept
+        {
+            m_coroutine->m_handle = nullptr;
+            if (m_awaiter)
+                m_awaiter.resume();
+            return {};
+        }
+
+        std::coroutine_handle<> m_awaiter;
+        Coroutine* m_coroutine = nullptr;
+    };
+
+    struct CoroutinePromiseValue : CoroutinePromiseBase {
+        template<typename U>
+        void return_value(U&& returned_object)
+        {
+            new (this->m_coroutine->return_value()) T(forward<U>(returned_object));
+        }
+
+        void return_value(T&& returned_object)
+        {
+            new (this->m_coroutine->return_value()) T(move(returned_object));
+        }
+    };
+
+    struct CoroutinePromiseVoid : CoroutinePromiseBase {
+        void return_void() { }
+    };
+
+    Coroutine(std::coroutine_handle<promise_type>&& handle)
+        : m_handle(move(handle))
+    {
+        m_handle.promise().m_coroutine = this;
+    }
+
+    T* return_value()
+    {
+        return reinterpret_cast<T*>(this->m_return_value);
+    }
+
+    std::coroutine_handle<promise_type> m_handle;
+};
+
+namespace Detail {
+template<typename T>
+struct TryAwaiter {
```

Comment:
Why is CO_TRY preferable to making ErrorOr awaitable? i.e. `co_await TRY(...);` or... no that doesn't work. `co_await ErrorOr { TRY(...) }`. Nope that doesn't work either. 

I guess I've got mixed feelings about two things:

- Why is this awaiter hidden in Detail, and not a handy dandy type we want to spell everywhere?
- Why do we need another macro to wrap expressions in?

Reply:
> Why is this awaiter hidden in Detail, and not a handy dandy type we want to spell everywhere?

Because of the same reason why we have `TRY` macro and not spell `if (result.is_error()) { return result.release_error(); }` everywhere like in go.

> Why do we need another macro to wrap expressions in?

Because `return` inside `TRY` won't work in coroutines.

I could have done
```cpp
#define CO_TRY(expression)                                                                           \
    ({                                                                                               \
        /* Ignore -Wshadow to allow nesting the macro. */                                            \
        AK_IGNORE_DIAGNOSTIC("-Wshadow",                                                             \
            auto&& _temporary_result = (expression));                                                \
        static_assert(!::AK::Detail::IsLvalueReference<decltype(_temporary_result.release_value())>, \
            "Do not return a reference from a fallible expression");                                 \
        if (_temporary_result.is_error()) [[unlikely]]                                               \
            co_return _temporary_result.release_error();                                             \
        _temporary_result.release_value();                                                           \
    })
```
but this awaiter always does 1 less move.